### ISSUE TITLE
java.net.URI is not implimented

### ISSUE THREAD

**Issue by aoiroaoino** [scala-native/scala-native#558]
ref. https://github.com/julien-truffaut/Monocle/pull/508
```
> show coreNative/nativeMissingDependencies
[info] * @java.net.URI
[info] * @java.net.URI::init_class.java.lang.String
[info] * @java.net.URI::toString_class.java.lang.String
```

**Comment by sjrd** [comment]
We have [an implementation in Scala.js](https://github.com/scala-js/scala-js/blob/master/javalib/src/main/scala/java/net/URI.scala) but it uses regular expressions.

**Comment by densh** [comment]
@aoiroaoino Thanks for report!

**Comment by densh** [comment]
Now that regex has been merged, we can port the Scala.js' version of URI.

**Comment by asoltysik** [comment]
Is somebody working on this?

**Comment by MasseGuillaume** [comment]
@asoltysik all yours!

**Comment by asoltysik** [comment]
Scala.js uses Javascript regex engine. I have issues trying to get URI regex to work with our re2 implementation. My proposition is to use a RFC 3986 compliant parser, like this one: http://uriparser.sourceforge.net/

What's your opinion about that? Should be try to debug Scala.js' regex or should we use the mentioned parser?

**Comment by asoltysik** [comment]
Scala.js implementation uses lookaheads which re2 does not support: https://github.com/scala-js/scala-js/blob/master/javalib/src/main/scala/java/net/URI.scala#L691

Any idea how to replace lookaheads here?

**Comment by MasseGuillaume** [comment]
I think you should write a parser. A regex for something complex like URI is a code smell in my opinion.

take a look at https://mathiasbynens.be/demo/url-regex.

One of them goes like this:

```
_(^|[\s.:;?\-\]<\(])(https?://[-\w;/?:@&=+$\|\_.!~*\|'()\[\]%#,☺]+[\w/#](\(\))?)(?=$|[\s',\|\(\).:;?\-\[\]>\)])_i
```

Compare it to this parser in akka-http:
https://github.com/akka/akka-http/blob/master/akka-http-core/src/main/scala/akka/http/impl/model/parser/UriParser.scala#L106

```scala
def URI = rule { scheme ~ ':' ~ `hier-part` ~ optional('?' ~ rawQueryString) ~ optional('#' ~ fragment) }
```

**Comment by asoltysik** [comment]
The problem here is that the parser in akka-http uses parboiled, we don't have a parser combinator here, so it won't look so pretty (unless we write our own). I'll start writing my own recursive descent parser though. At first glance it shouldn't be much of trouble.

**Comment by densh** [comment]
@asoltysik Recursive descent is probably the way to go here. 👍

**Comment by densh** [comment]
Fixed in #759
